app/services/depreciation_calculator.rb

```diff
@@ -0,0 +1,59 @@
+class DepreciationCalculator
+
+  def initialize(financial_year_reference, depreciation_period)
+    @fy_reference = financial_year_reference
+    @depreciation_period = depreciation_period
+  end
+
+  # We base the default start day of the financial year on the day following the reference FinancialYear as
+  # it can have an length that is not 1 year
+  def financial_year_start_day
+    @fy_reference.stopped_on + 1.day
+  end
+
+  def depreciation_period(started_on, yearly_percentage)
+    months = monthly_periods(started_on, yearly_percentage)
+
+    return months if @depreciation_period == :monthly
+
+    fy_start = financial_year_start_day
+    fy_shift = (fy_start - fy_start.beginning_of_year).days
+
+    groups = case @depreciation_period
+             when :quarterly
+               months.group_by do |m|
+                 shifted = m.started_on - fy_shift
+                 [shifted.year, shifted.month % 3]
+               end
+             when :yearly
+               months.group_by { |m_started_on, _, _| (m_started_on - fy_shift).year }
+             else
+               return nil
+             end
+
+    groups.map { |_, v| ([v.first.first, v.last.second, v.map(&:third).reduce(&:+)]) }
+  end
+
+  def monthly_periods(started_on, yearly_percentage)
+    return if(yearly_percentage == 0)
+
+    depreciation_duration = 100 / yearly_percentage * 360 # in days
```

Comment:
This will be automatically rounded if `yearly_percentage` happens to have been passed as an integer. Since we call it form the model for now it's not an issue but I'd feel better if we made it explicitly Float with a `to_f`.